README.md
# MDAT

## Medical Decision Aid Tool

[![Build Status](https://travis-ci.org/ctsit/mdat.svg?branch=master)](https://travis-ci.org/ctsit/mdat)
[![Build Status](https://travis-ci.org/ctsit/mdat.svg?branch=develop)](https://travis-ci.org/ctsit/mdat)

The MDAT is a set of software libraries to implement a _medical decision aid_.  These libraries are designed to select the best of two or more alternatives given responses to a list of criteria.  

## Installation

To install from source, use setup.py

    python setup.py install

## Usage Instructions

_mdat --help_ provides current usage instructions:
 
    usage: mdat [-h] [-i {json}] [-o {json,brief,csv,csvnoheader}]
                [infile] [outfile]
    
    Select the best of two or more alternatives given responses to a list of
    criteria
    
    positional arguments:
      infile
      outfile
    
    optional arguments:
      -h, --help            show this help message and exit
      -i {json}, --input {json}
                            Specify the file type used as input. Valid types: json
      -o {json,brief,csv,csvnoheader}, --output {json,brief,csv,csvnoheader}
                            Specify the file type used as input. Valid types:
                            json, brief, csv, csvnoheader

## Input data

Sample input data should be a json file with two or more labeled alternatives, one or more labeled criteria, and the numeric responses to each criteria for each alternative.  This example input file shows response data for 4 criteria that could be used to determine the preferred type of colorectal cancer screening:

    {
        "Accuracy": {
            "Fecal Immunochemical Test": 0.1,
            "Flexible Sigmoidoscopy": 0.2,
            "Colonoscopy": 0.3
        },
        "Amount of colon examined": {
            "Fecal Immunochemical Test": 0.5,
            "Flexible Sigmoidoscopy": 0.7,
            "Colonoscopy": 0.8
        },
        "Complications": {
            "Fecal Immunochemical Test": 0.2,
            "Flexible Sigmoidoscopy": 0.1,
            "Colonoscopy": 0.9
        },
        "Cost": {
            "Fecal Immunochemical Test": 0.4,
            "Flexible Sigmoidoscopy": 0.8,
            "Colonoscopy": 0.9
        },
        "Discomfort": {
            "Fecal Immunochemical Test": 0.1,
            "Flexible Sigmoidoscopy": 0.3,
            "Colonoscopy": 0.9
        },
        "Sedation": {
            "Fecal Immunochemical Test": 1.0,
            "Flexible Sigmoidoscopy": 0.8,
            "Colonoscopy": 0.9
        },
        "Test Preparation": {
            "Fecal Immunochemical Test": 0.1,
            "Flexible Sigmoidoscopy": 0.2,
            "Colonoscopy": 1.0
        }
    }

## Output data

mdat can output data in three formats.

    $ mdat sample_data/test.json 
    Colonoscopy
    
    $ mdat -o csv sample_data/test.json 
    Fecal Immunochemical Test,Flexible Sigmoidoscopy,Colonoscopy
    0.4403085839715581,0.4472662851726272,0.887231735365237
    
    $ mdat -o json sample_data/test.json 
    {"best_alternative": "Colonoscopy", "choquet_scores": {"Fecal Immunochemical Test": 0.43414030720240016, "Flexible Sigmoidoscopy": 0.4729342460226981, "Colonoscopy": 0.9068722937866821}}

## Requirements

This project requires Python 2.7 or greater.
 
This project uses Travis CI for automated testing. Please revise the .travis.yml to reflect any changes in required libraries as changes are made to the software.

## Contributions

The MDAT Team welcomes contributions to this project. Please fork and send pull requests with your revisions. Any code changes must be accompanied by corresponding unit tests to be accepted. Please configure your account at https://travis-ci.org to test your commits to your fork so you get quick feedback on any issues created by your changes.   


CHANGELOG.md
# Change Log
All notable changes to the MDAT project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [0.3.0] - 2015-06-24
### Added
- Add setup.py and revise README.md to match (Philip Chase)
- Add main function and test_data.json to assist in testing (Philip Chase)
- Write BestAlternative class (Philip Chase and Alex Loiacano)

### Changed
- Relocate test data to 'sample_data' folder and add test data sets (Philip Chase)
- Refactor to simplify calling on command line (Philip Chase)
- Add 'jsonschema' to modules installed by travis during testing (Philip Chase)


## [0.2.0] - 2015-06-09
### Changed
- Turned mdat into a package (Alex Loiacono)
- Rename project from Mefdas to mdat (Philip Chase)
- Move unit tests into tests/ folder (Philip Chase)

### Added
- Add travis CI testing (Philip Chase)
- Stub out new class BestAlternative and provide example JSON input code (Philip Chase)
- Add ChoquetIntegral class
- Add whiteboard image of a fuzzy measure calculation with 3-4 criteria (Philip Chase)
- Add whiteboard image of shapley value calculation (Philip Chase)

## [0.1.1] - 2015-05-22
### Changed
- Added project name to readme

## [0.1.0] - 2015-05-22
### Added

- Initial release
- Add FuzzyMeasure class to compute a fuzzy measure given a dictionary of criteria and values


test_fuzzyMeasure.py
from unittest import TestCase
import mefdas

__author__ = 'pbc'


class TestFuzzyMeasure(TestCase):
    def test_init(self):
        # self.list_of_members = frozenset([])
        fm = mefdas.FuzzyMeasure()
        self.assertEqual(len(fm.list_of_members), 0)

        fm = mefdas.FuzzyMeasure()
        self.assertEqual(len(fm.list_of_members), 0)

        criteria = {'c1': .9, 'c2': 1, 'c3': .6}
        fm = mefdas.FuzzyMeasure(criteria)
        self.assertEqual(len(fm.list_of_members),3)

    def test_store_criteria(self):
        fm = mefdas.FuzzyMeasure()
        criteria = {'c1': .9, 'c2': 1, 'c3': .6}
        fm.store_criteria(criteria)
        criteria_labels = set(['c1','c2','c3'])
        self.assertSetEqual(fm.list_of_members,criteria_labels)
        self.assertDictEqual(fm.criteria,criteria)

    def test_make_all_subsets(self):
        list_of_members = set([])
        fm = mefdas.FuzzyMeasure()
        fm.list_of_members = frozenset(['a','b','c','d','e','f','g','h','i','j','k','l','m'])
        fm.make_all_subsets()
        self.assertEqual(len(fm.set_of_all_subsets), 2**len(fm.list_of_members))

    def test_set_fm_for_trivial_cases(self):
        fm = mefdas.FuzzyMeasure()
        fm.list_of_members = frozenset([1,2,3,4,5])
        fm.make_all_subsets()
        fm.set_fm_for_trivial_cases()
        self.assertEqual(fm.mu[frozenset([])],0)
        self.assertEqual(fm.mu[fm.list_of_members],1)

    def test_set_fm_for_singleton_sets(self):
        # initialize FuzzyMeasure instance
        fm = mefdas.FuzzyMeasure()
        criteria = {'c1': .9, 'c2': 1, 'c3': .6}
        fm.store_criteria(criteria)
        fm.make_all_subsets()
        fm.set_fm_for_trivial_cases()

        # test function of interest
        fm.set_fm_for_singleton_sets()
        self.assertEqual(fm.mu[frozenset(['c1'])],0.36)
        self.assertEqual(fm.mu[frozenset(['c2'])],0.4)
        self.assertEqual(fm.mu[frozenset(['c3'])],0.24)

    def test_set_fm_for_complex_sets(self):
        # initialize FuzzyMeasure instance
        criteria = {'c1': .9, 'c2': .8, 'c3': .6, 'c4': .2}
        fm = mefdas.FuzzyMeasure(criteria)
        #print fm.mu

        self.assertEqual(1, 1)


test_bestAlternative.py
from unittest import TestCase
import mefdas

__author__ = 'pbc'


class TestBestAlternative(TestCase):
    def test_get_criteria(self):
        self.assertEqual(1, 1)

    def test_sum_of_criteria_values(self):
        self.assertEqual(1, 1)

    def test_get_alternatives(self):
        self.assertEqual(1, 1)

    def test_get_values_for_an_alternative(self):
        self.assertEqual(1, 1)

    def test_calculate(self):
        self.assertEqual(1, 1)


test_choquetIntegral.py
from unittest import TestCase
import mefdas

__author__ = 'pbc'


class TestChoquetIntegral(TestCase):
  def test_get_criteria_keys_sorted_by_value(self):
    criteria = {'c1': .6, 'c2': .8, 'c3': .9, 'c4': .2}
    expected_key_order = ['c4', 'c1', 'c2', 'c3']
    ci = mefdas.ChoquetIntegral(criteria=criteria)
    ci.get_criteria_keys_sorted_by_value()
    # verify the correct attribute is set
    self.assertEqual(ci.criteria_keys_sorted_by_value, expected_key_order)
    # verify the correct array is returned is set
    self.assertEqual(ci.get_criteria_keys_sorted_by_value(), expected_key_order)

  def test_calculate(self):
    criteria = {'c1': .6, 'c2': .8, 'c3': .9, 'c4': .2}
    fuzzyMeasure = {
      frozenset(['c3', 'c4']): 0.32,
      frozenset(['c3', 'c2', 'c4']): 0.56,
      frozenset(['c3', 'c1']): 0.98,
      frozenset(['c3', 'c2']): 0.27,
      frozenset([]): 0,
      frozenset(['c3', 'c1', 'c4']): 0.98,
      frozenset(['c2']): 0.2375,
      frozenset(['c3', 'c2', 'c1', 'c4']): 1,
      frozenset(['c3', 'c2', 'c1']): 0.99,
      frozenset(['c1']): 0.325,
      frozenset(['c2', 'c1']): 0.81,
      frozenset(['c2', 'c1', 'c4']): 0.99,
      frozenset(['c4']): 0.25,
      frozenset(['c1', 'c4']): 0.97,
      frozenset(['c2', 'c4']): 0.27,
      frozenset(['c3']): 0.1875
    }
    expected_ChoquetIntegral = 0.66875
    ci = mefdas.ChoquetIntegral(criteria=criteria, fuzzyMeasure=fuzzyMeasure)
    ci.get_criteria_keys_sorted_by_value()
    self.assertAlmostEquals(ci.calculate(), expected_ChoquetIntegral)


mdat.py

#!/usr/bin/env python

"""mdat.py: medical decision aid tool"""

__author__ = "Philip Chase(pbc@ufl.edu, Chris Barnes(cpb@ufl.edu), " \
             "Roy Keyes (keyes@ufl.edu), Alex Loiacono (atloiaco@ufl.edu)"
__copyright__ = "Copyright 2015, CTS-IT University of Florida"

import itertools
import random
import operator
import json

def make_all_subsets(list_of_members):
        # make every possible subsets of given list_of_members
        # for size in (list_of_members):
            # use combinations to enumerate all combinations of size elements
            # append all combinations to self.data

        set_of_all_subsets = set([])

        for i in range(len(list_of_members),-1,-1):
            for element in itertools.combinations(list_of_members,i):
                set_of_all_subsets.add(frozenset(element))

        return sorted(set_of_all_subsets)

class FuzzyMeasure:
    '''A class to produce a fuzzy measure of based on a list of criteria'''

    def __init__(self, criteria={}):
        # initialize a class to hold all fuzzyMeasure related objects
        self.set_of_all_subsets = set([])
        self.mu = ()
        self.criteria = criteria
        self.list_of_members = frozenset(criteria.keys())
        self.setup()

    def setup(self):
        if len(self.criteria) < 1 :
            return
        self.make_all_subsets()
        self.set_fm_for_trivial_cases()
        self.set_fm_for_singleton_sets()
        self.set_fm_for_complex_sets()

    def store_criteria(self, criteria):
        self.list_of_members = frozenset(criteria.keys())
        self.criteria = criteria

    def make_all_subsets(self):
        # make every possible subsets of given list_of_members
        self.set_of_all_subsets = make_all_subsets(self.list_of_members)

    def set_fm_for_trivial_cases(self):
        # set fuzzyMeasure for empty and complete sets
        # mu[] := 0
        # mu[X] := 1
        self.mu = {frozenset(): 0, self.list_of_members: 1}

    def set_fm_for_singleton_sets(self):
        '''set fuzzyMeasure for sets with exactly one member'''

        sum_of_criteria_values = 0
        for criterium in self.criteria:
            sum_of_criteria_values += self.criteria[criterium]

        for criterium in self.criteria:
            singleton_set = frozenset([criterium])
            self.mu[singleton_set] = self.criteria[criterium]/sum_of_criteria_values

    def set_fm_for_complex_sets(self):
        # set fuzzyMeasure for sets with 2 or more members

        # Random generation of a fuzzy measure mu on a set X
        # note: 'undefined' means we have not yet calculated and stored the value of mu for mu(foo)
        # create list of sets from X and shuffle the list
        list_of_all_subsets = list(self.set_of_all_subsets)
        random.shuffle(list_of_all_subsets)

        for A in list_of_all_subsets:
            if self.mu.get(A) is None:
                minimum_for_mu_A = 0
                maximum_for_mu_A = 1

                subsets_of_A = make_all_subsets(A)

                for B in subsets_of_A:
                    if self.mu.get(B) is not None:
                        minimum_for_mu_A = max(self.mu.get(B), minimum_for_mu_A)

                for B in list_of_all_subsets:
                    if self.mu.get(B) is not None:
                        if B.issuperset(A):
                            maximum_for_mu_A = min(maximum_for_mu_A, self.mu.get(B))

                self.mu[A] = random.uniform(minimum_for_mu_A,maximum_for_mu_A)

class ChoquetIntegral:
    ''' A class to calculate the Choquet Integral given a dictionary of criteria and a related fuzzy measure.

        Inputs
        ------
        The keys in the dictionary of criteria must match the names of the set members used to generate
        the keys to the fuzzy measure values. e.g. the keys in these criteria

            {
                'c1': 0.2,
                'c2': 0.4,
                'c3': 0.1
            }

        could be used to generate the frozensets that are used as the key values to look up values in this dictionary
        of fuzzy measures

            {
                frozenset([]): 0,
                frozenset(['c1']): 0.391304347826087,
                frozenset(['c2']): 0.3478260869565218,
                frozenset(['c3']): 0.2608695652173913,
                frozenset(['c1', 'c2']): 0.7683779330072605,
                frozenset(['c1', 'c3']): 0.8093446720056068,
                frozenset(['c2', 'c3']): 0.41548536225285937,
                frozenset(['c1', 'c2', 'c3']): 1
            }

        Outputs
        ------
        The class returns a single real number that is the Choquet Integral.  This value is the sum of the
        products of the differences between sorted criteria values and the corresponding fuzzy measures.

    '''

    def __init__(self,criteria={}, fuzzyMeasure={}):
        '''instantiate a Choquet Integral object'''
        self.criteria = criteria
        self.mu = fuzzyMeasure

    def get_criteria_keys_sorted_by_value(self):
        '''create the attribute, sigma, a list of criteria sorted by value'''
        self.criteria_keys_sorted_by_value = []

        # sort the criteria by value
        criteria_sorted_by_value = sorted(self.criteria.items(), key=operator.itemgetter(1))

        # make a list of the criteria keys in the
        for criterium in criteria_sorted_by_value:
            self.criteria_keys_sorted_by_value.append(criterium[0])

        return self.criteria_keys_sorted_by_value

    def calculate(self):
        '''Calculate the Choquet Integral and return just that value'''

        # initialize variables for loop
        self.utility=0
        x_n_minus_1 = 0
        self.get_criteria_keys_sorted_by_value()
        my_keys = self.criteria_keys_sorted_by_value[:]
        set_of_criteria = frozenset(my_keys)

        for criterum in self.criteria_keys_sorted_by_value:
            self.utility += (self.criteria[criterum] - x_n_minus_1) * self.mu[set_of_criteria]
            # set up for next loop
            x_n_minus_1 = self.criteria[criterum]
            my_keys.pop(0)
            set_of_criteria = frozenset(my_keys)

        return self.utility

class BestAlternative:
    ''' A class to determine the best alternative given a matrix of labeled alternatives and labeled criteria.

        Input
        -----

        Input data can be in the form of a python dictionary or JSON string with this structure:

        {
            "scores": {
                "accuracy": {
                    "fit": "0.1",
                    "sig": "0.2",
                    "col": "0.3"
                },
                "comfort": {
                    "fit": "0.1",
                    "sig": "0.2",
                    "col": "0.3"
                },
                "duration": {
                    "fit": "0.1",
                    "sig": "0.2",
                    "col": "0.3"
                },
                "time": {
                    "fit": "0.1",
                    "sig": "0.2",
                    "col": "0.3"
                }
            }
        }

        CSV should also be determined, but that input format is to be determined.

        Output
        ------

        Output should be dict or json string with this structure:

        {
            "best_alternative": "fit",
            "details": {
                "choquet_scores": {
                    "fit": "2.8",
                    "sig": "1.2",
                    "col": "2.0"
                },
                "library_version" : "1.0.3",
                ...
            }
        }

        As much or as few details as desired is acceptable. Only "best_alternative" is required.

    '''

    def __init__(self, scores={}, jsonScores='', csvScores=''):
        '''Initialize BestAlternative instance and store input data.
        Input must be the array of responses to N labeled criteria about M labeled alternatives.  '''
        if len(scores) > 0:
            self.scores=scores
            return()
        if len(jsonScores) > 0:
            self.scores=json.loads(jsonScores)
            return()
        if len(csvScores) > 0:
            print("error: csv input is not yet supported")
            return(2)
        else:
            print("error: No input supplied")
            return(3)

    def get_criteria(self):
        '''return a list containing labels for each criterium'''
        pass

    def sum_of_criteria_values(self):
        '''return a dictionary of the sum of each criterium across the alternatives keyed on the criteria labels'''
        pass

    def get_alternatives(self):
        '''return a list containing the labels for each alternative'''
        pass

    def get_values_for_an_alternative(self):
        '''return a dictionary of values for a single alternative, keyed on criterium'''
        pass

    def calculate(self):
        pass
        # get_criteria from input
        # compute sum_of_criteria_values based on input
        # calculate fuzzyMeasure
        #
        # get_alternatives from input
        # get_values_for_an_alternative from input
        # for alternative in list_of_alternatives:
        #     calculate ChoquetIntegral
        #     store ChoquetIntegral keyed on alternative_label
        #
        # return alternative_label for the highest stored choquetIntegral


if __name__ == "__main__":
    import sys
    mdat(int(sys.argv[1]))

mdat/__init__.py

#!/usr/bin/env python

"""mdat.py: medical decision aid tool"""

__author__ = "Philip Chase(pbc@ufl.edu, Chris Barnes(cpb@ufl.edu), " \
             "Roy Keyes (keyes@ufl.edu), Alex Loiacono (atloiaco@ufl.edu)"
__copyright__ = "Copyright 2015, CTS-IT University of Florida"

import itertools
import random
import operator
import json
import jsonschema
import sys
import argparse

def make_all_subsets(list_of_members):
        # make every possible subsets of given list_of_members
        # for size in (list_of_members):
            # use combinations to enumerate all combinations of size elements
            # append all combinations to self.data

        set_of_all_subsets = set([])

        for i in range(len(list_of_members),-1,-1):
            for element in itertools.combinations(list_of_members,i):
                set_of_all_subsets.add(frozenset(element))

        return sorted(set_of_all_subsets)

class FuzzyMeasure:
    '''A class to produce a fuzzy measure of based on a list of criteria'''

    def __init__(self, criteria={}):
        # initialize a class to hold all fuzzyMeasure related objects
        self.set_of_all_subsets = set([])
        self.mu = ()
        self.criteria = criteria
        self.list_of_members = frozenset(criteria.keys())
        self.setup()

    def setup(self):
        if len(self.criteria) < 1 :
            return
        self.make_all_subsets()
        self.set_fm_for_trivial_cases()
        self.set_fm_for_singleton_sets()
        self.set_fm_for_complex_sets()

    def store_criteria(self, criteria):
        self.list_of_members = frozenset(criteria.keys())
        self.criteria = criteria

    def make_all_subsets(self):
        # make every possible subsets of given list_of_members
        self.set_of_all_subsets = make_all_subsets(self.list_of_members)

    def set_fm_for_trivial_cases(self):
        # set fuzzyMeasure for empty and complete sets
        # mu[] := 0
        # mu[X] := 1
        self.mu = {frozenset(): 0, self.list_of_members: 1}

    def set_fm_for_singleton_sets(self):
        '''set fuzzyMeasure for sets with exactly one member'''

        sum_of_criteria_values = 0
        for criterium in self.criteria:
            sum_of_criteria_values += self.criteria[criterium]

        for criterium in self.criteria:
            singleton_set = frozenset([criterium])
            self.mu[singleton_set] = self.criteria[criterium]/sum_of_criteria_values

    def set_fm_for_complex_sets(self):
        # set fuzzyMeasure for sets with 2 or more members

        # Random generation of a fuzzy measure mu on a set X
        # note: 'undefined' means we have not yet calculated and stored the value of mu for mu(foo)
        # create list of sets from X and shuffle the list
        list_of_all_subsets = list(self.set_of_all_subsets)
        random.shuffle(list_of_all_subsets)

        for A in list_of_all_subsets:
            if self.mu.get(A) is None:
                minimum_for_mu_A = 0
                maximum_for_mu_A = 1

                subsets_of_A = make_all_subsets(A)

                for B in subsets_of_A:
                    if self.mu.get(B) is not None:
                        minimum_for_mu_A = max(self.mu.get(B), minimum_for_mu_A)

                for B in list_of_all_subsets:
                    if self.mu.get(B) is not None:
                        if B.issuperset(A):
                            maximum_for_mu_A = min(maximum_for_mu_A, self.mu.get(B))

                self.mu[A] = random.uniform(minimum_for_mu_A,maximum_for_mu_A)

class ChoquetIntegral:
    ''' A class to calculate the Choquet Integral given a dictionary of criteria and a related fuzzy measure.

        Inputs
        ------
        The keys in the dictionary of criteria must match the names of the set members used to generate
        the keys to the fuzzy measure values. e.g. the keys in these criteria

            {
                'c1': 0.2,
                'c2': 0.4,
                'c3': 0.1
            }

        could be used to generate the frozensets that are used as the key values to look up values in this dictionary
        of fuzzy measures

            {
                frozenset([]): 0,
                frozenset(['c1']): 0.391304347826087,
                frozenset(['c2']): 0.3478260869565218,
                frozenset(['c3']): 0.2608695652173913,
                frozenset(['c1', 'c2']): 0.7683779330072605,
                frozenset(['c1', 'c3']): 0.8093446720056068,
                frozenset(['c2', 'c3']): 0.41548536225285937,
                frozenset(['c1', 'c2', 'c3']): 1
            }

        Outputs
        ------
        The class returns a single real number that is the Choquet Integral.  This value is the sum of the
        products of the differences between sorted criteria values and the corresponding fuzzy measures.

    '''

    def __init__(self,criteria={}, fuzzyMeasure={}):
        '''instantiate a Choquet Integral object'''
        self.criteria = criteria
        self.mu = fuzzyMeasure

    def get_criteria_keys_sorted_by_value(self):
        '''create the attribute, sigma, a list of criteria sorted by value'''
        self.criteria_keys_sorted_by_value = []

        # sort the criteria by value
        criteria_sorted_by_value = sorted(self.criteria.items(), key=operator.itemgetter(1))

        # make a list of the criteria keys in the
        for criterium in criteria_sorted_by_value:
            self.criteria_keys_sorted_by_value.append(criterium[0])

        return self.criteria_keys_sorted_by_value

    def calculate(self):
        '''Calculate the Choquet Integral and return just that value'''

        # initialize variables for loop
        self.utility=0
        x_n_minus_1 = 0
        self.get_criteria_keys_sorted_by_value()
        my_keys = self.criteria_keys_sorted_by_value[:]
        set_of_criteria = frozenset(my_keys)

        for criterum in self.criteria_keys_sorted_by_value:
            self.utility += (self.criteria[criterum] - x_n_minus_1) * self.mu[set_of_criteria]
            # set up for next loop
            x_n_minus_1 = self.criteria[criterum]
            my_keys.pop(0)
            set_of_criteria = frozenset(my_keys)

        return self.utility

class BestAlternative:
    ''' A class to determine the best alternative given a matrix of labeled alternatives and labeled criteria.

        Input
        -----

        Input data can be in the form of a python dictionary or JSON string with this structure:

            {
                "accuracy": {
                    "fit": 0.1,
                    "sig": 0.2,
                    "col": 0.3
                },
                "comfort": {
                    "fit": 0.4,
                    "sig": 0.5,
                    "col": 0.6
                },
                "duration": {
                    "fit": 0.7,
                    "sig": 0.8,
                    "col": 0.9
                },
                "time": {
                    "fit": 0.4,
                    "sig": 0.3,
                    "col": 0.2
                }
            }

        CSV should also be determined, but that input format is to be determined.

        Output
        ------

        Output should be dict with this structure:

        {
            "best_alternative": "fit",
            "choquet_scores": {
                "fit": 2.8,
                "sig": 1.2,
                "col": 2.0
            },
            "library_version" : "1.0.3",
            ...
        }

        Only "best_alternative" is required.  Other output is optional and expandable.

    '''

    def __init__(self, scores={}, jsonScores='', csvScores=''):
        '''Initialize BestAlternative instance and store input data.
        Input must be the array of responses to N labeled criteria about M labeled alternatives.  '''
        if len(scores) > 0:
            self.setup(json.loads(json.dumps(scores)))
            self.scores=scores
            return(None)
        if len(jsonScores) > 0:
            self.setup(json.loads(jsonScores))
            self.scores=json.loads(jsonScores)
            return(None)
        if len(csvScores) > 0:
            print("error: csv input is not yet supported")
            return(None)
        else:
            print("error: No input supplied")
            return(None)

    def setup(self, jsonScores):
        '''verify the jsonScores is structured correctly'''
        schema = '''
            {
                "$schema": "http://json-schema.org/draft-04/schema#",
                "title": "mdat input",
                "description": "An input data set for the Medical Decision Aids Tool python library",
                "type": "object",
                "patternProperties": {
                    "^.+$": {
                        "$ref": "#definitions/alternatives"
                    }
                },
                "minProperties": 1,
                "definitions": {
                    "alternatives": {
                        "title": "alternative",
                        "description": "an alternative to choose from",
                        "type": "object",
                        "patternProperties": {
                            "^.+$": {
                                "type": "number",
                                "minimum" : 0,
                                "maximum" : 1
                            }
                        },
                        "minProperties": 2
                    }
                }
            }
        '''
        jsonschema.validate(jsonScores, json.loads(schema))

    def get_criteria(self):
        '''return a list containing labels for each criterium'''
        list_of_criteria = self.scores.keys()

    def sum_of_criteria_values(self):
        '''return a dictionary of the sum of each criterium across the alternatives keyed on the criteria labels'''
        dict_of_sums = {}
        for criteria in self.list_of_criteria:
            dict_of_sums[criteria] = sum(float,self.scores[criteria].values())

    def get_alternatives(self):
        '''return a list containing the labels for each alternative'''
        pass

    def get_values_for_an_alternative(self):
        '''return a dictionary of values for a single alternative, keyed on criterium'''
        pass

    def calculate(self):
        # get_criteria from input
        # compute sum_of_criteria_values based on input
        # calculate fuzzyMeasure
        #
        # get_alternatives from input
        # get_values_for_an_alternative from input
        # for alternative in list_of_alternatives:
        #     calculate ChoquetIntegral
        #     store ChoquetIntegral keyed on alternative_label
        #
        # return alternative_label for the highest stored choquetIntegral
        resultJSONString = '''
            {
                "best_alternative": "fit",
                "choquet_scores": {
                    "fit": 2.8,
                    "sig": 1.2,
                    "col": 2.0
                }
            }
        '''
        result = json.loads(resultJSONString)
        return(result)


def main():
    '''
    A class to determine the best alternative given a matrix of labeled alternatives and labeled criteria.

    Input
    -----

    Input data can be in the form of a JSON string with this structure:

        {
            "accuracy": {
                "fit": 0.1,
                "sig": 0.2,
                "col": 0.3
            },
            "comfort": {
                "fit": 0.4,
                "sig": 0.5,
                "col": 0.6
            },
            "duration": {
                "fit": 0.7,
                "sig": 0.8,
                "col": 0.9
            },
            "time": {
                "fit": 0.4,
                "sig": 0.3,
                "col": 0.2
            }
        }


    Output
    ------

    Output data is either a string with the label of the best alternative or a json string in this format

    {
        "best_alternative": "fit",
        "choquet_scores": {
            "fit": 2.8,
            "sig": 1.2,
            "col": 2.0
        }
    }


    Usage:

        usage: mdat.py [-h] [-i {json}] [-o {json,brief}] [infile] [outfile]

        Select the best of two or more alternatives given responses to a list of
        criteria

        positional arguments:
          infile
          outfile

        optional arguments:
          -h, --help            show this help message and exit
          -i {json}, --input {json}
                                Specify the file type used as input. Valid types: json
          -o {json,brief}, --output {json,brief}
                                Specify the file type used as input. Valid types:
                                json, text
    '''

    # define the list of acceptable arguments
    parser = argparse.ArgumentParser(
        description='Select the best of two or more alternatives given responses to a list of criteria')
    parser.add_argument('infile', nargs='?', type=argparse.FileType('r'), default=sys.stdin)
    parser.add_argument('outfile', nargs='?', type=argparse.FileType('w'), default=sys.stdout)
    parser.add_argument(
        '-i',
        '--input',
        choices=['json'],
        dest='input',
        default='json',
        help='Specify the file type used as input. Valid types: json')
    parser.add_argument(
        '-o',
        '--output',
        choices=['json', 'brief'],
        dest='output',
        default='brief',
        help='Specify the file type used as input. Valid types: json, text')

    # prepare the arguments we were given
    args = parser.parse_args()

    # Prepare the input
    if args.input == 'json':
        ba = BestAlternative(jsonScores=args.infile.read())
    else:
        print "Unsupported input type"
        return()

    # generate and return the output
    if args.output == 'json':
        args.outfile.write(json.dumps(ba.calculate()))
    elif args.output == 'brief':
        args.outfile.write(ba.calculate()['best_alternative'])
    else:
        print "Unsupported output type"
        return()


if __name__ == "__main__":
    main()
